pkg/vector/math.go

```diff
@@ -0,0 +1,52 @@
+package vector
+
+import (
+	"math"
+
+	"go.rtnl.ai/nlp/pkg/errors"
+)
+
+// Cosine returns the cosine of the angle between two vectors; which can be used
+// as a similarity metric (as defined by SLP 3rd Edition section 6.4 fig 6.10).
+// If the vectors do not have the same number of elements, an error will be
+// returned.
+func Cosine(a, b Vector) (cosine float64, err error) {
+	// Ensure vectors have the same number of elements
+	if len(a) != len(b) {
+		return 0.0, errors.ErrUnequalLengthVectors
+	}
+
+	var (
+		dp, vla, vlb float64
+	)
+	if dp, err = DotProduct(a, b); err != nil {
+		return 0.0, err
+	}
+	vla = VectorLength(a)
+	vlb = VectorLength(b)
```

Comment:
We should add a check here to prevent a divide by zero error when either of the vector lengths are 0.